### ISSUE TITLE
swapchain creation on second screen fails

### ISSUE THREAD

**Issue by jojo61** [haasn/libplacebo#57]
When I try to create a swapchain with a surface from another screen eg.g.. :0.1  then this fails
Here is the debug output:
debug: Requested image count: 2 (min 2 max 8)
 info: (Re)creating swapchain of size 1920x1080
error: vkCreateSwapchainKHR(vk->dev, &sinfo, VK_ALLOC, &p->swapchain): VK_ERROR_INITIALIZATION_FAILED
error: Failed (re)creating swapchain!
Is this a shortcoming from vulkan or is it a problem in libplacebo ?  The surfacecreation itself is successfull.

// create XCB surface for swapchain	
	xcbinfo.sType	= VK_STRUCTURE_TYPE_XCB_SURFACE_CREATE_INFO_KHR;
	xcbinfo.pNext	= NULL;
	xcbinfo.flags	= 0;
	xcbinfo.connection = Connection;
	xcbinfo.window = VideoWindow;

	if (vkCreateXcbSurfaceKHR(p->vk_inst->instance,&xcbinfo,NULL,&p->pSurface) != VK_SUCCESS) {
		Fatal(_("Failed to create XCB Surface\n"));
	}

**Comment by haasn** [comment]
Try turning on validation (`vk_inst_params.debug = true`)?

What is meant by "second screen" in this context? Is this about multiple displays attached to the same X server? Don't those use the same socket (i.e. `:0`)?

**Comment by haasn** [comment]
Also, do you assign your `VkSurfaceKHR` to `pl_vulkan_params.surface` when creating the `pl_vulkan` device? Because if not, try doing that.

**Comment by jojo61** [comment]
Yes it is a second screen attached to the same X Server. And yes I assign the Surface also to  pl_vulkan_params.surface when creating the vulkan device.  I will provide a trace with debug by tomorrow.

One side notice. Even when pl_vulkan_create_swapchain fails it will return a swapchain pointer. So it is difficult to see that it failed. It should return NULL in that case.

**Comment by haasn** [comment]
> One side notice. Even when pl_vulkan_create_swapchain fails it will return a swapchain pointer. So it is difficult to see that it failed. It should return NULL in that case.

Good point. We can easily change that now, to make sure the swapchain actually *works*. Fixed in adab9b7.

> Yes it is a second screen attached to the same X Server. And yes I assign the Surface also to pl_vulkan_params.surface when creating the vulkan device. I will provide a trace with debug by tomorrow.

Please do that. To be honest it might also be a driver bug, especially if you're on older drivers. In theory we check to make sure the swapchain format etc. are supported by the physical device, and if you also pass the surface when creating the device then we check for compatibility at that stage as well. So as far as the vulkan API is concerned, creating the swapchain *should* work.

**Comment by haasn** [comment]
@jojo61 I need the full log, not just the last few lines. And put the verbosity to `PL_LOG_TRACE`

What I'm confused about is why DISPLAY works the way it does for you. I'm testing on a machine with two screens now and DISPLAY=:0.1 doesn't work at all (in any X-using program). Do you somehow have two X servers running on different displays?

Anyway, I'm not sure if there's anything I can do here, at least not from the information provided. Does `DISPLAY=:0.1 demos/sdl2` work? (After building that demo, see the instructions at the top of the file)

>  vdr205: ../src/ta/ta.c:336: ta_dbg_check_header: Assertion `h->canary == CANARY' failed.

Fixed that error.

**Comment by jojo61** [comment]
Ok I will provide full log by tomorrow. 
I have only one X Server with 2 Displays configured in xorg.conf. 
With your latest source I have problems on Display :0.1 as well. Some of your changes made things worse. I will investigate further and come back to you.

**Comment by haasn** [comment]
> Some of your changes made things worse. I will investigate further and come back to you.

I noticed some regressions myself caused by 52c21bb (and also adab9b7). If the window gets resized in between `swap_buffers` and `frame_start`, the resize won't get picked up by the latter due to API limitations.

**Comment by jojo61** [comment]
Attached is a full trace of the failed Init Sequence. At the end of the trace I have included the sourcepart of the Init Sequence. 
[trace.txt.gz](https://github.com/haasn/libplacebo/files/2610124/trace.txt.gz)

**Comment by jojo61** [comment]
I read somewhere: 
With Xcb, minImageExtent, maxImageExtent, and currentExtent are the window size. Therefore, a swapchain’s imageExtent must match the window’s size. 
But before creating the swapchain I never give the windowsize down to placebo.  Is there a API shortcoming for xcb surfaces ?

**Comment by haasn** [comment]
> But before creating the swapchain I never give the windowsize down to placebo. Is there a API shortcoming for xcb surfaces ?

We query the current window size via the vulkan API. (We set the swapchain's imageExtent to the currentExtent, which, as you said, is the window size).

Thanks for the log. Unfortunately, I don't see anything wrong in it either from our side or from yours. Somebody more familiar with the nvidia driver would have to look into it. (Which is, sadly, closed source - so I can't investigate this further myself)

FWIW, I decided to revert 52c21bb and adab9b7 on master because they were causing issues with images being acquired before a resize operation (but rendered to after it). Unfortunately, this also means that your swapchain creation will continue to fail.

Another way of resolving those issues relies on the extension `VK_KHR_shared_presentable_image`, which is not supported in any common driver so I can't test the code. (But I pushed it to branch swapchain_status nonetheless)
